coap/api/transport.go

```diff
@@ -0,0 +1,206 @@
+	return func(conn *net.UDPConn, addr *net.UDPAddr, msg *gocoap.Message) *gocoap.Message {
+		var res *gocoap.Message
+		if msg.IsConfirmable() {
+			res = &gocoap.Message{
+				Type:      gocoap.Acknowledgement,
+				Code:      gocoap.Content,
+				MessageID: msg.MessageID,
+				Token:     msg.Token,
+				Payload:   []byte{},
+			}
+			res.SetOption(gocoap.ContentFormat, gocoap.AppJSON)
+		}
+
+		if len(msg.Payload) == 0 && msg.IsConfirmable() {
+			res.Code = gocoap.BadRequest
+			return res
+		}
+
+		cid := mux.Var(msg, "id")
+		publisher, err := coap.Authorize(msg, res, cid)
+		if err != nil {
+			res.Code = gocoap.Unauthorized
+			return res
+		}
+
+		rawMsg := mainflux.RawMessage{
+			Channel:   cid,
+			Publisher: publisher,
+			Protocol:  protocol,
+			Payload:   msg.Payload,
+		}
+
+		if err := svc.Publish(rawMsg); err != nil {
+			if msg.IsConfirmable() {
+				res.Code = gocoap.InternalServerError
+			}
+			return res
+		}
+
+		if msg.IsConfirmable() {
+			res.Code = gocoap.Changed
+		}
+		return res
+	}
+}
+
+func observe(svc coap.Service) func(conn *net.UDPConn, addr *net.UDPAddr, msg *gocoap.Message) *gocoap.Message {
+	return func(conn *net.UDPConn, addr *net.UDPAddr, msg *gocoap.Message) *gocoap.Message {
+		var res *gocoap.Message
+		if msg.IsConfirmable() {
+			res = &gocoap.Message{
+				Type:      gocoap.Acknowledgement,
+				Code:      gocoap.Content,
+				MessageID: msg.MessageID,
+				Token:     msg.Token,
+				Payload:   []byte{},
+			}
+			res.SetOption(gocoap.ContentFormat, gocoap.AppJSON)
+		}
+
+		cid := mux.Var(msg, "id")
+		publisher, err := coap.Authorize(msg, res, cid)
+
+		if err != nil {
+			res.Code = gocoap.Unauthorized
+			return res
+		}
+
+		if value, ok := msg.Option(gocoap.Observe).(uint32); ok && value == 1 {
+			id := fmt.Sprintf("%s-%x", publisher, msg.Token)
+			err := svc.Unsubscribe(id)
+			if err != nil {
+				res.Code = gocoap.InternalServerError
+			}
+		}
+
+		if value, ok := msg.Option(gocoap.Observe).(uint32); ok && value == 0 {
+			ch := make(chan mainflux.RawMessage)
+			id := fmt.Sprintf("%s-%x", publisher, msg.Token)
+			if err := svc.Subscribe(cid, id, ch); err != nil {
+				res.Code = gocoap.InternalServerError
+				return res
+			}
+			go handleSub(svc, id, conn, addr, msg, ch)
+			res.AddOption(gocoap.Observe, 0)
+		}
+		return res
+	}
+}
+
+func notify(svc coap.Service, id string, conn *net.UDPConn, addr *net.UDPAddr, msg *gocoap.Message, counter *uint16) error {
+	err := sendMessage(conn, addr, msg, counter)
+	if err != nil {
+		return svc.Unsubscribe(id)
+	}
+	return nil
+}
+
+func sendMessage(conn *net.UDPConn, addr *net.UDPAddr, msg *gocoap.Message, counter *uint16) error {
+	var err error
+	*counter++
+	now := time.Now().UnixNano() / int64(time.Millisecond)
+	buff := new(bytes.Buffer)
+	err = binary.Write(buff, binary.LittleEndian, now)
+	if err != nil {
+		return err
+	}
+	msg.MessageID = *counter
+	msg.SetOption(gocoap.Observe, buff.Bytes()[:3])
+	for i := 0; i < 3; i++ {
```

Comment:
Where does this 3 times comes from? Spec?

Reply:
No, as I know, protocol specification is not explicit when it comes to this number, as well as timeout between confirmable message and corresponding acknowledgement or reject message.